README.md
This is a simple mirror of traceroute releases.
Upstream author: Dmitry Butskoy

Upstream homepage:
http://traceroute.sourceforge.net/

Released tarballs:
https://sourceforge.net/projects/traceroute/files/traceroute/

I'd appreciate if the upstream author Dmitry Butskoy publish any repository (doesn't have to be git :-)).


ChangeLog
2016-02-15  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.22

	*  Some portability fixing and improvements (Felix Janda)

	*  Require clear numbers for options and arguments (Sergey Salnikov)

	*  Drop compilation date from the version string (Debian #774365)

	*  New tcp module option `reuse', which utilize SO_REUSEADDR
	   to reuse local port numbers for the huge workloads (Richard Sheehan)

	*  Avoid poll(2) call with spurious zero timeout in some rare cases
	   by rounding the value properly using ceil(3)


2014-11-12  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.21

	*  Fix `--mtu' and `-F' working on kernels >= 3.13

	*  Some manual page improving (Christopher Mann)


2014-06-14  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.20

	*  Describe all complementary long options in the man page (Jan Synacek)

	*  Use correct service name for AS lookups (Frederic Mangano)

	*  Avoid some rare case null dereference (geogriffin@jsgriff.com)

	*  Improve expiration check for simultaneous probes


2012-11-19  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.19

	*  DCCP protocol support (rfc4340), by Samuel Jero

	   Use "-D" option for it (the protocol-specific options
	   are available too).

	*  Update COPYING and COPYING.LIB license files to the latest
	   published ones (due to FSF address changes etc.) (Jan Synacek)

	*  Add mention of "-l" option to manual (Filip Holec)


2011-08-16  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.18

	*  Handle new dgram icmp sockets ("echo ping sockets"),
	   appeared in kernel 3.0 .

	   Now unprivileged users may perform ICMP tracerouting
	   without any special rights of the executable
	   (neither setuid bits nor cap_net_raw settings).
	   It is allowed if any group of a user matches sysctl range
	   of "net/ipv4/ping_group_range".

	   The support for dgram icmp way (and whether it is allowed)
	   is auto-detected at runtime. First, the traditional raw socket
	   is tried (for full compatibility reasons), then new dgram
	   socket as a fallback.
	   The icmp module now has two additional options "raw" and "dgram",
	   which cause to try one particular way only.

	   Note, that there is no IPv6 implementation for dgram icmp sockets
	   in kernels 3.0 yet, but new traceroute is ready for it anyway.

	*  New tcp module option `info' ("-T -O info"),
	   which prints all tcp flags of tcp reply from the reached
	   target host.

	   The flags are shown comma-separated in the same place
	   where icmp extensions is printed (ie. in `<>' brackets)

	   This feature is utilized by tcptraceroute wrapper now,
	   and allow it to be completely functional replacement
	   of the original tcptraceroute.

	*  Fix determination of system-wide ECN setings for tcp module.

	   Since the kernel 2.6.31 the default sysctl net/ipv4/tcp_ecn
	   was changed from zero to '2', whereas the actual value
	   for ecn to be set is still '1'

	*  Allow different packet sizes for `--mtu'.
	   Suport `-l' option for tracepath wrapper.

	*  Some code and manual cleanups


2010-12-14  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.17

	*  Adapt code to make possible the use of Linux capabilities
	   (for raw sockets etc.) instead of superuser privileges only.

	   On modern systems the capabilities can be stored as
	   file attributes, ie.:

	   "setcap cap_net_raw=pe /usr/bin/traceroute"


2010-09-13  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.16

	*  A little work-around in the build system
	   for the new (buggy?) make 3.82

	*  Add `--fwmark=num' option for firewall mark (for kernel >= 2.6.25).
	   Idea comes from an anonymous SF patch #3042539


2010-07-14  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.15

	*  Use string routines more safely (fix SF bug #3029216)

	*  Provide help for lft wrapper


2010-04-21  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.14

	*  Fix support for IPv6's flow_labels and tclass.
	   Thanks to Peter Bieringer for testing

	*  Use route header "type 2" instead of deprecated "type 0"
	   for `-g' option for IPv6. The default value can be changed
	   by specifying a number in the place of the first `-g' address.


2009-11-02  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.13

	*  Check for first_hop is not zero value (vladz@devzero.fr)

	*  Always fill unresolved IP address by its numeric interpretation,
	   even if getnameinfo(3) leaves it untouched (as it does for ipv6
	   in some glibc versions, whereas always fills for ipv4)

	*  Cosmetic changes for man page (Andreas Mohr)


2008-09-15  Dmitry Butskoy  <Dmitry@Butskoy.name> - 2.0.12

	*  Use common recv_reply() routine for all modules which
	   do recvmsg(2) call. Method-specific things go to callbacks.

	   Pass to init methods pointer to datalen instead of the value.

	*  Implement ICMP Extension support (rfc4884), `-e' option.
	   Parse MPLS info (rfc4950) to be more readable (Kaj Niemi)

	*  Implement Path MTU Discovery (similar to tracepath(1)),
	   with `--mtu' option. Changed mtu is printed once in a form 
	   of `F=NUM' at the first probe of a hop which requires
	   such mtu to be reached. (Actually, the correspond "frag needed"
	   icmp message is normally sent by the previous hop).

	*  Print the number of backward hops when it differs with forward,
	   by `--back' option. The backward hops is guessed by a technique
	   similar to tracepath(1), there is no reliable way to obtain
	   such info though.

	*  The optional second argument (packet_len) now is the full length
	   of the packet, including IP headers. (It is obvious enough due to
	   the nature of this feature, and this is the behaviour of the
	   original traceroute). Particular trace methods can ignore this
	   (fe. tcp), or increase it up to the minimal value (udp, icmp).
	   The actual packet's size is alvays reported in the output header.

	*  Add tracepath(1)/tracepath6(1) shell wrapper.

	*  Allow DEF_AF to be redefined at cmdline (Teran McKinney)

	*  Do not check the correctness of `sim_probes' value -- it is

traceroute/udp.c
/*
    Copyright (c)  2006		    Dmitry K. Butskoy
				    <buc@citadel.stu.neva.ru>
    License:  GPL		

    See COPYING for the status of this software.
*/

#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <time.h>
#include <sys/time.h>
#include <netinet/icmp6.h>
#include <netinet/ip_icmp.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <errno.h>
#include <linux/types.h>
#include <linux/errqueue.h>

#include "traceroute.h"


static sockaddr_any dest_addr = {{ 0, }, };
static unsigned short dest_port = DEF_UDP_PORT;

static char *data;
static size_t data_len = 0;


static int udp_init (const char *host, const sockaddr_any *dest,
				unsigned int port_seq, size_t packet_len) {
	int i;

	dest_addr = *dest;
	dest_addr.sin.sin_port = 0;

	if (port_seq)  dest_port = port_seq;

	data_len = packet_len;
	data = malloc (data_len);
	if (!data)  error ("malloc");

        for (i = 0; i < data_len; i++)
                data[i] = 0x40 + (i & 0x3f);
 
	return 0;
}


static void udp_send_probe (probe *pb, int ttl) {
	int sk, val = 0;
	int af = dest_addr.sa.sa_family;


	sk = socket (af, SOCK_DGRAM, 0);
	if (sk < 0)  error ("socket");

	tune_socket (sk);	/*  common stuff   */

	if (af == AF_INET) {

	    val = ttl;
	    if (setsockopt (sk, SOL_IP, IP_TTL, &val, sizeof (val)) < 0)
		    error ("setsockopt IP_TTL");

	    val = 1;
	    if (setsockopt (sk, SOL_IP, IP_RECVERR, &val, sizeof (val)) < 0)
		    error ("setsockopt IP_RECVERR");

	}
	else if (af == AF_INET6) {

	    val = ttl;
	    if (setsockopt (sk, SOL_IPV6, IPV6_UNICAST_HOPS,
						&val, sizeof (val)) < 0
	    )  error ("setsockopt IPV6_UNICAST_HOPS");

	    val = 1;
	    if (setsockopt (sk, SOL_IPV6, IPV6_RECVERR, &val, sizeof (val)) < 0)
		    error ("setsockopt IPV6_RECVERR");
	}


	pb->send_time = get_time ();

	dest_addr.sin.sin_port = htons (dest_port);	/* both ipv4 and ipv6 */
	dest_port++;

	if (connect (sk, &dest_addr.sa, sizeof (dest_addr)) < 0)
		error ("connect");

	if (send (sk, data, data_len, 0) < 0)
		error ("send");


	pb->sk = sk;

	add_poll (sk, POLLERR);

	return;
}


static void udp_recv_probe (int sk, int revents, probe *probes,
						unsigned int num_probes) {
	struct msghdr msg;
	sockaddr_any from;
	char control[1024];
	struct cmsghdr *cm;
	int i;
	probe *pb;


	if (!(revents | POLLERR))
		return;


	memset (&msg, 0, sizeof (msg));

	msg.msg_name = &from;
	msg.msg_namelen = sizeof (from);
	msg.msg_control = control;
	msg.msg_controllen = sizeof (control);

	if (recvmsg (sk, &msg, MSG_ERRQUEUE) < 0)
		return;


	for (i = 0; i < num_probes && probes[i].sk != sk; i++) ;
	if (i >= num_probes)  return;
	pb = &probes[i];


	for (cm = CMSG_FIRSTHDR (&msg); cm; cm = CMSG_NXTHDR (&msg, cm)) {

	    if ((cm->cmsg_level == SOL_IP && cm->cmsg_type == IP_RECVERR) ||
		(cm->cmsg_level == SOL_IPV6 && cm->cmsg_type == IPV6_RECVERR)
	    ) {
		struct sock_extended_err *ee;

		ee = (struct sock_extended_err *) CMSG_DATA (cm);
		memcpy (&pb->res, SO_EE_OFFENDER (ee), sizeof (pb->res));

		if (ee->ee_origin == SO_EE_ORIGIN_ICMP ||
		    ee->ee_origin == SO_EE_ORIGIN_ICMP6
		)  parse_icmp_res (pb, ee->ee_type, ee->ee_code);
	    }
	}

	pb->recv_time = get_timestamp (&msg);

	pb->recv_ttl = get_recv_ttl (&msg);


	del_poll (sk);

	close (sk);
	pb->sk = -1;

	pb->done = 1;
}


static void udp_expire_probe (probe *pb) {

	del_poll (pb->sk);

	close (pb->sk);
	pb->sk = -1;

	pb->done = 1;
}


tr_ops udp_ops = {
	.init = udp_init,
	.send_probe = udp_send_probe,
	.recv_probe = udp_recv_probe,
	.expire_probe = udp_expire_probe,
};



traceroute/icmp.c
/*
    Copyright (c)  2006		    Dmitry K. Butskoy
				    <buc@citadel.stu.neva.ru>
    License:  GPL		

    See COPYING for the status of this software.
*/

#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/poll.h>
#include <time.h>
#include <sys/time.h>
#include <netinet/icmp6.h>
#include <netinet/ip_icmp.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <errno.h>
#include <linux/types.h>
#include <linux/errqueue.h>

#include "traceroute.h"


static sockaddr_any dest_addr = {{ 0, }, };
static u_int16_t seq = 1;
static u_int16_t ident = 0;

static char *data;
static size_t data_len = 0;

static int icmp_sk = -1;
static int last_ttl = 0;


static u_int16_t in_cksum (const void *ptr, size_t len) {
	const u_int16_t *p = (const u_int16_t *) ptr;
	unsigned int sum = 0;
	u_int16_t res;

	while (len > 1) {
	    sum += *p++;
	    len -= 2;
	}

	if (len)
	    sum += htons (*((unsigned char *) p) << 8);

	sum = (sum >> 16) + (sum & 0xffff);
	sum += (sum >> 16);
	res = ~sum;
	if (!res)  res = ~0;

	return res;
}


static int icmp_init (const char *host, const sockaddr_any *dest,
				unsigned int port_seq, size_t packet_len) {
	int i;
	int af = dest->sa.sa_family;

	dest_addr = *dest;
	dest_addr.sin.sin_port = 0;

	if (port_seq)  seq = port_seq;

	data_len = sizeof (struct icmphdr) + packet_len;
	data = malloc (data_len);
	if (!data)  error ("malloc");

        for (i = sizeof (struct icmphdr); i < data_len; i++)
                data[i] = 0x40 + (i & 0x3f);


	icmp_sk = socket (af, SOCK_RAW, (af == AF_INET) ? IPPROTO_ICMP
							: IPPROTO_ICMPV6);
	if (icmp_sk < 0)
		error ("socket");

	tune_socket (icmp_sk);

	add_poll (icmp_sk, POLLIN);

	ident = getpid () & 0xffff;
 
	return 0;
}


static void icmp_send_probe (probe *pb, int ttl) {
	int af = dest_addr.sa.sa_family;


	if (ttl != last_ttl) {

	    if (af == AF_INET) {
		if (setsockopt (icmp_sk, SOL_IP, IP_TTL, &ttl, sizeof(ttl)) < 0)
			error ("setsockopt IP_TTL");
	    }
	    else if (af == AF_INET6) {
		if (setsockopt (icmp_sk, SOL_IPV6, IPV6_UNICAST_HOPS,
						    &ttl, sizeof (ttl)) < 0
		)  error ("setsockopt IPV6_UNICAST_HOPS");
	    }

	    last_ttl = ttl;
	}


	if (af == AF_INET) {
	    struct icmp *icmp = (struct icmp *) data;

	    icmp->icmp_type = ICMP_ECHO;
	    icmp->icmp_code = 0;
	    icmp->icmp_cksum = 0;
	    icmp->icmp_id = htons (ident);
	    icmp->icmp_seq = htons (seq);

	    icmp->icmp_cksum = in_cksum (data, data_len);
	}
	else if (af == AF_INET6) {
	    struct icmp6_hdr *icmp6 = (struct icmp6_hdr *) data;

	    icmp6->icmp6_type = ICMP6_ECHO_REQUEST;
	    icmp6->icmp6_code = 0;
	    icmp6->icmp6_cksum = 0;
	    icmp6->icmp6_id = htons (ident);
	    icmp6->icmp6_seq = htons(seq);

	    icmp6->icmp6_cksum = in_cksum (data, data_len);
	}


	pb->send_time = get_time ();

	if (sendto (icmp_sk, data, data_len, 0,
			    &dest_addr.sa, sizeof (dest_addr)) < 0
	)  error ("sendto");


	pb->seq = seq;

	seq++;

	return;
}


static void icmp_recv_probe (int sk, int revents, probe *probes,
						unsigned int num_probes) {
	int af = dest_addr.sa.sa_family;
	struct msghdr msg;
	sockaddr_any from;
	struct iovec iov;
	int n, type, code;
	u_int16_t recv_id, recv_seq;
	probe *pb;
	char buf[1024];		/*  enough, enough...  */
	char control[1024];


	if (!(revents | POLLIN))
		return;

	memset (&msg, 0, sizeof (msg));
	msg.msg_name = &from;
	msg.msg_namelen = sizeof (from);
	msg.msg_control = control;
	msg.msg_controllen = sizeof (control);
	iov.iov_base = buf;
	iov.iov_len = sizeof (buf);
	msg.msg_iov = &iov;
	msg.msg_iovlen = 1;

	n = recvmsg (sk, &msg, 0);
	if (n < sizeof (struct icmphdr))	/*  error or too short   */
		return;


	if (af == AF_INET) {
	    struct iphdr *ip = (struct iphdr *) buf;
	    int hlen = ip->ihl << 2;
	    struct icmp *icmp;

	    n -= hlen + sizeof (struct icmphdr);
	    if (n < 0)  return;

	    icmp = (struct icmp *) (buf + hlen);
	    type = icmp->icmp_type;
	    code = icmp->icmp_code;

	    if (type == ICMP_ECHOREPLY) {
		    recv_id = ntohs (icmp->icmp_id);
		    recv_seq = ntohs (icmp->icmp_seq);
	    }
	    else if (type == ICMP_TIME_EXCEEDED ||
		     type == ICMP_DEST_UNREACH
	    ) {
		if (n < sizeof (struct iphdr) + sizeof (struct icmphdr))
			return;

		ip = (struct iphdr *) (((char *)icmp) + sizeof(struct icmphdr));
		hlen = ip->ihl << 2;

		if (n < hlen + sizeof (struct icmphdr))
			return;
		if (ip->protocol != IPPROTO_ICMP)
			return;

		icmp = (struct icmp *) (((char *) ip) + hlen);
		recv_id = ntohs (icmp->icmp_id);
		recv_seq = ntohs (icmp->icmp_seq);

	    } else
		return;
	}
	else {	    /*  AF_INET6   */
	    struct icmp6_hdr *icmp6 = (struct icmp6_hdr *) buf;

	    type = icmp6->icmp6_type;
	    code = icmp6->icmp6_code;

	    if (type == ICMP6_ECHO_REPLY) {
		    recv_id = ntohs (icmp6->icmp6_id);
		    recv_seq = ntohs (icmp6->icmp6_seq);
	    }
	    else if (type == ICMP6_TIME_EXCEEDED ||
		     type == ICMP6_DST_UNREACH
	    ) {
		struct ip6_hdr *ip6;

		if (n < 2 * sizeof (struct icmp6_hdr) + sizeof (struct ip6_hdr))
			return;

		ip6 = (struct ip6_hdr *) (icmp6 + 1);
		if (ip6->ip6_nxt != IPPROTO_ICMPV6)
			return;

		icmp6 = (struct icmp6_hdr *) (ip6 + 1);
		recv_id = ntohs (icmp6->icmp6_id);
		recv_seq = ntohs (icmp6->icmp6_seq);

	    } else
		return;
	}


	if (recv_id != ident)
		return;

	for (n = 0; n < num_probes && probes[n].seq != recv_seq; n++) ;
	if (n >= num_probes)  return;
	pb = &probes[n];


	memcpy (&pb->res, &from, sizeof (pb->res));

	if ((af == AF_INET && type == ICMP_ECHOREPLY) ||
	    (af == AF_INET6 && type == ICMP6_ECHO_REPLY)
	) {
	    pb->final = 1;
	} else
	    parse_icmp_res (pb, type, code);

	pb->recv_time = get_timestamp (&msg);

	pb->recv_ttl = get_recv_ttl (&msg);


	pb->seq = -1;

	pb->done = 1;
}


static void icmp_expire_probe (probe *pb) {

	pb->seq = -1;

	pb->done = 1;
}


tr_ops icmp_ops = {
	.init = icmp_init,
	.send_probe = icmp_send_probe,
	.recv_probe = icmp_recv_probe,
	.expire_probe = icmp_expire_probe,
};


traceroute/traceroute.h
/*
    Copyright (c)  2006		    Dmitry K. Butskoy
				    <buc@citadel.stu.neva.ru>
    License:  GPL		

    See COPYING for the status of this software.
*/


union common_sockaddr {
	struct sockaddr sa;
	struct sockaddr_in sin;
	struct sockaddr_in6 sin6;
};
typedef union common_sockaddr sockaddr_any;

struct probe_struct {
	int done;
	int final;
	sockaddr_any res;
	double send_time;
	double recv_time;
	int recv_ttl;
	int sk;
	int seq;
	char err_str[16];	/*  assume enough   */
};
typedef struct probe_struct probe;

struct traceroute_operations {
	int (*init) (const char *host, const sockaddr_any *dest,
				unsigned int port_seq, size_t packet_len);
	void (*send_probe) (probe *pb, int ttl);
	void (*recv_probe) (int fd, int revents,
				probe *probes, unsigned int num_probes);
	void (*expire_probe) (probe *pb);
};
typedef struct traceroute_operations tr_ops;

extern tr_ops udp_ops, icmp_ops, tcp_ops;


#define DEF_UDP_PORT	33434
#define DEF_TCP_PORT	80

void error (const char *str) __attribute__((noreturn));

double get_time (void);
void tune_socket (int sk);
void parse_icmp_res (probe *pb, int type, int code);

void use_timestamp (int sk);
double get_timestamp (struct msghdr *msg);
void use_recv_ttl (int sk);
int get_recv_ttl (struct msghdr *msg);

void add_poll (int fd, int events);
void del_poll (int fd);

extern const char *get_as_path (const char *query);


traceroute/mod-dccp.c
/*
    Copyright (c)  2012		Samuel Jero <sj323707@ohio.edu>
    License:  GPL v2 or any later

    See COPYING for the status of this software.
*/

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <poll.h>
#include <netinet/ip.h>
#include <netinet/ip6.h>
#include <linux/dccp.h>


#include "traceroute.h"


#define DEF_SERVICE_CODE 	1885957735

#define DCCP_HEADER_LEN		(sizeof (struct dccp_hdr) + \
				 sizeof (struct dccp_hdr_ext) \
				 + sizeof (struct dccp_hdr_request))

static sockaddr_any dest_addr = {{ 0, }, };
static unsigned int dest_port = 0;

static int raw_sk = -1;
static int last_ttl = 0;

static uint8_t buf[1024];	/*  enough, enough...  */
static size_t csum_len = 0;
static struct dccp_hdr *dh = NULL;
static struct dccp_hdr_ext *dhe = NULL;
static struct dccp_hdr_request *dhr = NULL;
static unsigned int service_code = DEF_SERVICE_CODE;


static CLIF_option dccp_options[] = {
	{ 0, "service", "NUM", "Set DCCP service code to %s (default is "
				_TEXT (DEF_SERVICE_CODE) ")",
				CLIF_set_uint, &service_code, 0, CLIF_ABBREV },
	CLIF_END_OPTION
};


static int dccp_init (const sockaddr_any *dest,
				unsigned int port_seq, size_t *packet_len_p) {
	int af = dest->sa.sa_family;
	sockaddr_any src;
	socklen_t len;
	uint8_t *ptr;
	uint16_t *lenp;


	dest_addr = *dest;
	dest_addr.sin.sin_port = 0;	/*  raw sockets can be confused   */

	if (!port_seq)  port_seq = DEF_DCCP_PORT;
	dest_port = htons (port_seq);


	/*  Create raw socket for DCCP   */
	raw_sk = socket (af, SOCK_RAW, IPPROTO_DCCP);
	if (raw_sk < 0)
		error_or_perm ("socket");

	tune_socket (raw_sk);	    /*  including bind, if any   */

	if (connect (raw_sk, &dest_addr.sa, sizeof (dest_addr)) < 0)
		error ("connect");

	len = sizeof (src);
	if (getsockname (raw_sk, &src.sa, &len) < 0)
		error ("getsockname");


	if (!raw_can_connect ()) {	/*  work-around for buggy kernels  */
		close (raw_sk);
		raw_sk = socket (af, SOCK_RAW, IPPROTO_DCCP);
		if (raw_sk < 0)  error ("socket");
		tune_socket (raw_sk);
		/*  but do not connect it...  */
	}


	use_recverr (raw_sk);

	add_poll (raw_sk, POLLIN | POLLERR);


	/*  Now create the sample packet.  */

	/*  For easy checksum computing:
		saddr
		daddr
		length
		protocol
		dccphdr
	*/

	ptr = buf;

	if (af == AF_INET) {
		len = sizeof (src.sin.sin_addr);
		memcpy (ptr, &src.sin.sin_addr, len);
		ptr += len;
		memcpy (ptr, &dest_addr.sin.sin_addr, len);
		ptr += len;
	} else {
		len = sizeof (src.sin6.sin6_addr);
		memcpy (ptr, &src.sin6.sin6_addr, len);
		ptr += len;
		memcpy (ptr, &dest_addr.sin6.sin6_addr, len);
		ptr += len;
	}

	lenp = (uint16_t *) ptr;
	ptr += sizeof (uint16_t);
	*((uint16_t *) ptr) = htons ((uint16_t) IPPROTO_DCCP);
	ptr += sizeof (uint16_t);


	/*  Construct DCCP header   */
	dh = (struct dccp_hdr *) ptr;

	dh->dccph_ccval = 0;
	dh->dccph_checksum = 0;
	dh->dccph_cscov = 0;
	dh->dccph_dport = dest_port;
	dh->dccph_reserved = 0;
	dh->dccph_sport = 0;	/*  temporary   */
	dh->dccph_x = 1;
	dh->dccph_type = DCCP_PKT_REQUEST;
	dh->dccph_seq2 = 0;	/*  reserved if using 48 bit sequence numbers  */
	/*  high 16 bits of sequence number. Always make 0 for simplicity.  */
	dh->dccph_seq = 0;
	ptr += sizeof (struct dccp_hdr);

	dhe = (struct dccp_hdr_ext *) ptr;
	dhe->dccph_seq_low = 0;		/*  temporary   */
	ptr += sizeof (struct dccp_hdr_ext);

	dhr = (struct dccp_hdr_request *) ptr;
	dhr->dccph_req_service = htonl (service_code);
	ptr += sizeof (struct dccp_hdr_request);


	csum_len = ptr - buf;

	if (csum_len > sizeof (buf))
		error ("impossible");	/*  paranoia   */

	len = ptr - (uint8_t *) dh;
	if (len & 0x03)  error ("impossible");  /*  as >>2 ...  */

	*lenp = htons (len);
	dh->dccph_doff = len >> 2;


	*packet_len_p = len;

	return 0;
}


static void dccp_send_probe (probe *pb, int ttl) {
	int sk;
	int af = dest_addr.sa.sa_family;
	sockaddr_any addr;
	socklen_t len = sizeof (addr);


	/*  To make sure we have chosen a free unused "source port",
	   just create, (auto)bind and hold a socket while the port is needed.
	*/

	sk = socket (af, SOCK_DCCP, IPPROTO_DCCP);
	if (sk < 0)  error ("socket");

	bind_socket (sk);

	if (getsockname (sk, &addr.sa, &len) < 0)
		error ("getsockname");

	/*  When we reach the target host, it can send us either Reset or Response.
	  For Reset all is OK (we and kernel just answer nothing), but
	  for Response we should reply with our Close.
	    It is well-known "half-open technique", used by port scanners etc.
	  This way we do not touch remote applications at all, unlike
	  the ordinary connect(2) call.
	    As the port-holding socket neither connect() nor listen(),
	  it means "no such port yet" for remote ends, and kernel always
	  send Reset in such a situation automatically (we have to do nothing).
	*/


	dh->dccph_sport = addr.sin.sin_port;

	dhe->dccph_seq_low = random_seq ();

	dh->dccph_checksum = 0;
	dh->dccph_checksum = in_csum (buf, csum_len);


	if (ttl != last_ttl) {
		set_ttl (raw_sk, ttl);
		last_ttl = ttl;
	}


	pb->send_time = get_time ();

	if (do_send (raw_sk, dh, dh->dccph_doff << 2, &dest_addr) < 0) {
		close (sk);
		pb->send_time = 0;
		return;
	}


	pb->seq = dh->dccph_sport;

	pb->sk = sk;

	return;
}


static probe *dccp_check_reply (int sk, int err, sockaddr_any *from,
						    char *buf, size_t len) {
	probe *pb;
	struct dccp_hdr *ndh = (struct dccp_hdr *) buf;
	uint16_t sport, dport;


	if (len < 8)  return NULL;	    /*  too short   */


	if (err) {
	    sport = ndh->dccph_sport;
	    dport = ndh->dccph_dport;
	} else {
	    sport = ndh->dccph_dport;
	    dport = ndh->dccph_sport;
	}


	if (dport != dest_port)
		return NULL;

	if (!equal_addr (&dest_addr, from))
		return NULL;

	pb = probe_by_seq (sport);
	if (!pb)  return NULL;

	if (!err)  pb->final = 1;

	return pb;
}


static void dccp_recv_probe (int sk, int revents) {

	if (!(revents & (POLLIN | POLLERR)))
		return;

	recv_reply (sk, !!(revents & POLLERR), dccp_check_reply);
}


static void dccp_expire_probe (probe *pb) {

	probe_done (pb);
}


static tr_module dccp_ops = {
	.name = "dccp",
	.init = dccp_init,
	.send_probe = dccp_send_probe,
	.recv_probe = dccp_recv_probe,
	.expire_probe = dccp_expire_probe,
	.one_per_time = 0,
	.options = dccp_options,
};

TR_MODULE (dccp_ops);
